File: JAVA prog/BinaryHeap.java
public class BinaryHeap {
    private int [] Heap;
    private int size;
    private int maxsize;
    private static final int front = 1;

    public BinaryHeap(int ms) {
        maxsize = ms;
        size = 0;
        Heap = new int[ms+1];
        Heap[0] = Integer.MIN_VALUE;
    }

    private int parent(int pos) {
        return pos/2;
    }

    private int leftChild(int pos) {
        return (2*pos);
    }

    private int rightChild(int pos) {
        return (2*pos)+1;
    }

    private boolean isLeaf(int pos) {
        if(pos >= (size/2) && pos <= size)
            return true;
        return false;
    }

    private void swap(int i, int j) {
        int t = Heap[i];
        Heap[i] = Heap[j];
        Heap[j] = t;
    }

    private void minHeapify(int pos) {
        if(!isLeaf(pos)) {
            if(Heap[pos] > Heap[leftChild(pos)] || Heap[pos] > Heap[rightChild(pos)]) {
                if(Heap[leftChild(pos)] < Heap[rightChild(pos)]) {
                    swap(pos,leftChild(pos));
                    minHeapify(leftChild(pos));
                }
                else {
                    swap(pos,rightChild(pos));
                    minHeapify(rightChild(pos));
                }
            }
        }
    }

    public void insert(int element) {
        if(size >= maxsize)
            return;
        Heap[++size] = element;
        int current = size;

        while (Heap[current] < Heap[parent(current)]) {
            swap(current, parent(current));
            current = parent(current);
        }
    }

    public void print()
    {
        for (int i = 1; i <= size / 2; i++) {
            System.out.print(" PARENT : " + Heap[i]
                    + " LEFT CHILD : " + Heap[2 * i]
                    + " RIGHT CHILD :" + Heap[2 * i + 1]);
            System.out.println();
        }
    }

    public void minHeap()
    {
        for (int pos = (size / 2); pos >= 1; pos--) {
            minHeapify(pos);
        }
    }

    public int remove()
    {
        int popped = Heap[front];
        Heap[front] = Heap[size--];
        minHeapify(front);
        return popped;
    }
}

File: code/data_structures/src/binary_heap/binary_heap.dart
class Node {
  int priority;
  dynamic value;

  Node(this.priority, this.value);
}

class BinaryHeap {
  late List<Node> heap;

  BinaryHeap() {
    heap = [];
  }

  void insert(int priority, dynamic value) {
    final newNode = Node(priority, value);
    heap.add(newNode);
    _bubbleUp(heap.length - 1);
  }

  Node extractMin() {
    if (heap.isEmpty) {
      throw Exception("Heap is empty");
    }

    final min = heap[0];
    final last = heap.removeLast();
    if (heap.isNotEmpty) {
      heap[0] = last;
      _heapify(0);
    }

    return min;
  }

  void _bubbleUp(int index) {
    while (index > 0) {
      final parentIndex = (index - 1) ~/ 2;
      if (heap[index].priority < heap[parentIndex].priority) {
        // Swap the elements if the current element has higher priority than its parent.
        final temp = heap[index];
        heap[index] = heap[parentIndex];
        heap[parentIndex] = temp;
        index = parentIndex;
      } else {
        break;
      }
    }
  }

  void _heapify(int index) {
    final leftChild = 2 * index + 1;
    final rightChild = 2 * index + 2;
    int smallest = index;

    if (leftChild < heap.length &&
        heap[leftChild].priority < heap[index].priority) {
      smallest = leftChild;
    }
    if (rightChild < heap.length &&
        heap[rightChild].priority < heap[smallest].priority) {
      smallest = rightChild;
    }

    if (smallest != index) {
      final temp = heap[index];
      heap[index] = heap[smallest];
      heap[smallest] = temp;
      _heapify(smallest);
    }
  }
}

void main() {
  final minHeap = BinaryHeap();

  minHeap.insert(4, "A");
  minHeap.insert(9, "B");
  minHeap.insert(2, "C");
  minHeap.insert(1, "D");
  minHeap.insert(7, "E");

  print("Extracted Min: ${minHeap.extractMin().value}"); // Should print "D"
  print("Extracted Min: ${minHeap.extractMin().value}"); // Should print "C"
  print("Extracted Min: ${minHeap.extractMin().value}"); // Should print "A"
}
